ldaptor/protocols/ldap/ldapsyntax.py

```diff
@@ -710,7 +710,7 @@ def search(self,
             cb = callback
         try:
             op = pureldap.LDAPSearchRequest(
-                baseObject=self.dn,
+                baseObject=self.dn.getText(),
```

Comment:
Originally there was an `str` interface:
```
baseObject=str(self.dn)
```
Later I changed it to passing just `DistinguishedName` instance. `LDAPSearchRequest` instance (as any other `pureber`/`pureldap` instances) use `to_bytes` function to get bytes representation of an object. It uses `toWire` method if it is defined for an object. But now there are no `toWire` methods in `distinguishedname` classes so I pass a unicode string.

I just thought that I can extend `to_bytes` function to use `getText` method for `distinguishedname` objects. So the interface will be the same and `to_bytes` will look something like this:
```
def to_bytes(value):
    if hasattr(value, 'toWire'):
        return value.toWire()
    if hasattr(value, 'getText'):
        return value.getText().encode('utf-8')
    if isinstance(value, six.text_type):
        return value.encode('utf-8')
    return bytes(value)
```

Reply:
If it was string/text interface. then is ok.

I don't like the updates in to_bytes :)
I think that an instance which does not have toWire should have be passed to to_bytes

If a DN needs to be serialized, we can add the toWire() method

Reply:
Good question :)

I am expecting `DistinguishedName` class to only have `getText` as in 90% of places where it is called you want the text representation.

so, 

I would have only `getText` and in that one place where it used to generate an LDAP we can manually call geText() and encode to ASN.1 strung

In https://www.ietf.org/rfc/rfc2251.txt Appendix A I see this definition.

```
Appendix A - Complete ASN.1 Definition

...some parts removed...

        LDAPString ::= OCTET STRING
        LDAPOID ::= OCTET STRING
        LDAPDN ::= LDAPString
        RelativeLDAPDN ::= LDAPString
```.

So, if would have a toWire method I was expecting it to output ASN.1 octet string not just the UTF-8 encoding.

So toWire() method is only used when constructing an LDAP message.

--------

Does it make sense?

Reply:
Hmm, yes, it became more clear. I found [this RFC section](https://tools.ietf.org/html/rfc4514#section-3) describing how DN is converted to ASN.1. I'd implement `toWire` method using this grammar so we can proceed with using this interface:
```
baseObject=self.dn
```
Would it be okay?

Sorry for a long answer and a long discussion of this PR. Didn't expect this issue to be such controversial :)

Reply:
Don't worry for the long answer.
I think that this is a good opportunity to understand the code.

My understanding is that `toWire` should only be used for the classes from 
https://github.com/twisted/ldaptor/blob/master/ldaptor/protocols/pureldap.py

these are the low-level classed dealing with the LDAP ASN.1 wire protocol (LDAP messages).

-------------

All the other classes are high level API or internal classes and we should not care about their wire representation.

I think that the classes from https://github.com/twisted/ldaptor/blob/master/ldaptor/protocols/ldap/distinguishedname.py should also go in pureldap.py as they don't have logic, but are rather the representation of some items/objects/entities in LDAP... and some of those items/objects/entities are hitting the wire.

----

Now the current code might not follow these rules...and it might be hard work to fix the API

So, what we can do now, is to make sure that we don't use `toWire` for something which does not has the semantic of "get me the lower level bytes LDAP message".

Anything that is a bit higher level and is not directly sent over the network, should have a different method name.

What do you think?

Reply:
Yep. It sounds good.

I prefer toWire() rather than toASN1, as I feel that it is more of a semantic naming.
I am ok with any name.

If it is easier for you, we can merge this as it is and do the follow up work in a separate PR.

Thanks!

Reply:
Great! I fixed other small issues, it can be merged now.